pkg/daemon/daemon.go

```diff
@@ -466,6 +477,16 @@ func (dn *Daemon) syncNode(key string) error {
 		return errors.Wrapf(err, "prepping update")
 	}
 	if current != nil || desired != nil {
+		// Only check for config drift if we need to update.
+		if err := dn.runPreflightConfigDriftCheck(); err != nil {
+			return err
+		}
+
+		// Make sure we don't attempt to validate on disk contents while the update
+		// is running
+		dn.diskValidatorLock.Lock()
+		defer dn.diskValidatorLock.Unlock()
+
 		if err := dn.triggerUpdateWithMachineConfig(current, desired); err != nil {
```

Comment:
This function gets called at multiple places. wondering if it will make more sense to do mutex lock inside triggerUpdateWithMachineConfig() function definition.

Reply:
I tested this and discovered that it can dead-lock under certain conditions. If we acquire the lock in `dn.triggerUpdateWithMachineConfig` and defer the unlock, we won't unlock until the call to `dn.update` returns.

When we call `dn.update`, we eventually get to `dn.performPostConfigChangeAction`. However, under certain conditions, `dn.performPostConfigChangeAction` will call [`dn.triggerUpdateWithMachineConfig` again](https://github.com/openshift/machine-config-operator/blob/master/pkg/daemon/update.go#L161).

However, since we haven't returned from the `dn.update` call in `dn.triggerUpdateWithMachineConfig`, the lock never gets released and it hangs on the second call waiting for the lock to become available.

Reply:
In the other places where `dn.triggerUpdateWithMachineConfig` is called, I don't think locking / unlocking matters as much because those are the booting or bootstrapping phases. Even though the periodic config drift checker is started by that time, that just starts the goroutine; it doesn't actually run a scan until after the check interval has been satisfied. I could add some guards to skip the scan if we're bootstrapping or booting but I don't think that's necessary.

Reply:
I added the guards in `dn.runPeriodicConfigDriftCheck` to check if we're booting or if we're bootstrapping. If either of those cases is met, we'll skip the config drift check.